src/client.rs

```diff
@@ -182,131 +280,141 @@ impl CoAPClient {
-                                warn!("reply ack failed {}", e)
-                            }
-                        }
+        tokio::spawn(async move {
+            let mut rx_pinned: Pin<
+                Box<
+                    dyn Future<
+                            Output = std::result::Result<ObserveMessage, oneshot::error::RecvError>,
+                        > + Send,
+                >,
+            > = Box::pin(rx);
+            loop {
+                tokio::select! {
+                    sock_rx = self.receive_from_socket() => {
+                        self.receive_and_handle_message(sock_rx, &mut handler).await;
                     }
+                     observe = &mut rx_pinned => {
+                    match observe {
+                    Ok(ObserveMessage::Terminate) => {
+                        let mut deregister_packet = CoapRequest::<SocketAddr>::new();
+                        deregister_packet.message.header.message_id =
+                            Self::gen_message_id(&mut message_id);
+                        deregister_packet.set_observe_flag(ObserveOption::Deregister);
+                        deregister_packet.set_path(observe_path.as_str());
+
+                        Self::send_with_socket(&self.transport, &deregister_packet.message)
+                            .await
+                            .unwrap();
+                        self.receive_from_socket().await.unwrap();
+                        break;
+                    }
+                    // if the receiver is dropped, we change the future to wait forever
+                    Err(_) => {debug!("observe continuing forever"); rx_pinned  = Box::pin(futures::future::pending()) },
                 }
-                Err(e) => match e.kind() {
-                    ErrorKind::WouldBlock => {
-                        info!("Observe timeout");
                     }
-                    _ => warn!("observe failed {:?}", e),
-                },
-            };
-
-            match observe_receiver.try_recv() {
-                Ok(ObserveMessage::Terminate) => {
-                    let mut deregister_packet = CoapRequest::<SocketAddr>::new();
-                    deregister_packet.message.header.message_id =
-                        Self::gen_message_id(&mut message_id);
-                    deregister_packet.set_observe_flag(ObserveOption::Deregister);
-                    deregister_packet.set_path(observe_path.as_str());
-
-                    Self::send_with_socket(&socket, &peer_addr, &deregister_packet.message)
-                        .unwrap();
-                    Self::receive_from_socket(&socket).unwrap();
-                    break;
+
                 }
-                _ => continue,
             }
         });
-        self.observe_sender = Some(observe_sender);
-        self.observe_thread = Some(observe_thread);
-
-        return Ok(());
-    }
-
-    /// Stop observing
-    pub fn unobserve(&mut self) {
```

Comment:
I changed 'observe' to take ownership of the client, I have renamed the termination logic and moved it to a separate function `terminate_observe` that should handle the logic for terminating.

Reply:
right now in master, trying to use the client after starting an observe can lead to a data race between the observe thread and the client.

Reply:
Cool. Can you add some codes about how to `unobserve` in the example `examples/client.rs` ? Because now it's need the caller use the oneshot sender to cancel observation.

Reply:
sure thing.